### DISCUSSION TITLE
Keycloak Authorisation with PKCE

### DISCUSSION THREAD

**Discussion by nickelmatt** [keycloak/keycloak#21103] [Q&A]
Hello,

I've got Keycloak's JS adapter to work with PKCE to get rid of sharing a secret key between the server and clients. However, to enable Authorisation, I need to change the access type to "confidential", which forces using one of 

- Signed JWT,
- client ID and secret,
- certificate,
- Signed JWT and secret key. 

The JS adapter doesn't seem to support any of these options, or am I missing something?

Maybe the better question is, should I look elsewhere for:

1. Authorisation services and/or
2. A JS adapter

**Comment by jonkoops** [comment]
We do have an [adapter for AuthZ](https://github.com/keycloak/keycloak/blob/main/js/libs/keycloak-js/src/keycloak-authz.js) that is based on Keycloak JS. I have not personally used this, but perhaps it can fit your use case?

**Reply by nickelmatt** [reply]
Thanks! I'll try it out and update this post if it works

**Reply by nickelmatt** [reply]
Update: I am still unable to proceed without including a secret key. The real problem is that the code will be deployed in a public environment and it's not recommended to use "confidential" access type, which is the only way to enable Authorisation. Thanks for your help anyway

**Reply by jonkoops** [reply]
Sorry to hear, I am not too familiar with the authz part of the code. But perhaps there is some way to accomplish this.

**Comment by fluke777** [comment]
Any progress, thoughts on this?

I am relatively new to Oauth but I share the same complications. Once I activate the authentication I start getting the 401 on token retrieval endpoint. 

Isn't PKCE whole reason to exist that I do not have to use client secret?
I can switch on authentication and authorization for the implicit flow.

I would like to at least understand if this is my misunderstanding of PKCE or a missing feature in Keycloak.
